### PR TITLE
Upgrade PyTorch

### PR DESCRIPTION
**Pull request by snarb**
Upgrade to PyTorch v 1.3
Changes:

- Removed outdated sync batch normalization. Unfortunately we still We cannot use official implementation of Sync BN: SyncBatchnorm when using nn.DataParallel, this work is in process AFIK on the torch team side. But you can use with some extra steps: https://github.com/dougsouza/pytorch-sync-batchnorm-example . I have one GPU so not using it. 
- Hardcoded Constants  numbers like image dimensions are moved to erf_settings.py. 
- prob_to_lines.py  - python implementation of matlab code.  Works in real time. 
- Import image is example of import custom dataset with different image detentions to the CUlane format. In my case  - this is lablebox.com format. Everything else is the same.  Also I have code with live demo on video, interpolation, outlier points filtering,  contact me if you are interesting.

I have achieved 0.60 mIoU on val CUlane dataset when training 14h+  from scratch on one 2080RTX GPU.  It is still slow improving, so I think it should be the same 0.61 as restrained model.

### CODE REVIEW

**Comment by TommyAnqi** [comment]
Hi, snarb, thanks for your sharing. I am really interested in your live demo , interpolation, and outlier points filtering. Could you share the code to me? My email is anqitommy@gmail.com. Thank you!

**Comment by JohnLiTangshan** [comment]
Hi @snarb  thanks for your sharing, I am interested in code for live demo on video, interpolation, outlier points filtering. Could you share the code with me? My email is 104479752@qq.com Thank you!

**Comment by SaeedAID** [comment]
@snarb Thanks for your great work! Please could you send me the code to saeed.turabzadeh@gmail.com?

**Comment by LukeAI** [comment]
@cardwing  would you consider merging this PR?

**Comment by cardwing** [comment]
I have not tested the uploaded code and I am not sure about its performance. So, if you have tested the code, please tell us its performance and then I will merge this PR.

**Comment by wzhang1** [comment]
Thank you for sharing, could you briefly describe the changes on erfnet.py, e.g. some of BatchNorm2d have track_running_stats=True while some does not, looks like pytorch 1.3 has this default true. Also in the traning code, any differece beween BCEWithLogitsLoss() and BCELoss()

**Comment by SaeedAID** [comment]
> * prob_to_lines.py  - python implementation of matlab code.  Works in real time.

@snarb could you send the real time code to saeed.turabzadeh@gmail.com or advise me on how to use prob_to_lines.py, please?

**Comment by yanconglin** [comment]
> I have not tested the uploaded code and I am not sure about its performance. So, if you have tested the code, please tell us its performance and then I will merge this PR.
Screenshot from 2020-02-02 16-48-19

Hi, @cardwing, I have tested the upgraded pytorch code (version 1.3). The training is with one gpu and takes nearly two days. The performance is more or less the same as yours (Fmeasure: 0.730205 in my case). Please check the attached picture for details. Once again, thanks @snarb for sharing the upgraded code.

**Comment by cardwing** [comment]
@yanconglin Thanks for your sharing. I will merge this PR right now.

**Comment by SaeedAID** [comment]
> * prob_to_lines.py  - python implementation of matlab code.  Works in real time.

@snarb I've tested your prob_to_lines.py  - python implementation of Matlab code, Thanks one more time for your time and effort. I found some bugs in it such, line 36 YS has not been defined also line 11 the linId calculation is different from the Matlab code and some more. I am just wondering that, how did you use the code? could you please share your thought on this python code, please? I've already asked you for the real time code that you did as I would like to test its performance in real-time such as frame rate and...

**Comment by snarb** [comment]
@SaeedAID yes, you are right, please add:  YS = IN_IMAGE_H - np.arange(POINTS_COUNT) * 20 - 1
there was little refactoring but results are the same as matlab code. I have made a real time demo with this code. Here is an example, it is composed from the real code on the fly so bugs are possible but I hope you will got the idea:


     while(True):
            check, in_frame = cap.read()
            im_frame = cv2.cvtColor(im_frame, cv2.COLOR_BGR2RGB)
            if not check:
                print("Last frame")
                break

           in_frame = cv2.resize(in_frame, (IN_IMAGE_W, IN_IMAGE_H), interpolation=cv2.INTER_LINEAR)

	    croppedImage = in_frame[VERTICAL_CROP_SIZE:, :, :]  # FIX IT
	    croppedImageTrain = cv2.resize(croppedImage, (TRAIN_IMG_W, TRAIN_IMG_H), interpolation=cv2.INTER_LINEAR)
	image = transform([croppedImageTrain.astype(np.float32), None])[0]
	 input_var = torch.autograd.Variable(image)
	output, output_exist = model(input_var)
	   output = F.softmax(output, dim=1)
	    pred = output.data.cpu().numpy()  # BxCxHxW
	    pred_exist = output_exist.data.cpu().numpy()  
	    maps = []
	    mapsResized = []
	    exists = []
	    for l in range(LANES_COUNT):
		prob_map = (pred[0][l + 1] * 255).astype(int)
		prob_map = cv2.blur(prob_map, (9, 9))
		prob_map = prob_map.astype(np.uint8)
		mapsResized.append(cv2.resize(prob_map, (IN_IMAGE_W, IN_IMAGE_H_AFTER_CROP), interpolation=cv2.INTER_LINEAR))
		curImg = AddMask(curImg, mask, COLORS[l]) # Image with probability map
		exists.append(pred_exist[0][l] > 0.5)
		lines = GetLines(exists, maps)

	...

	    for l in range(LANES_COUNT):
		  points = lines[l] # Points for the lane
	```

**Comment by fantasycheng** [comment]
Hi, snarb, thanks for your sharing. I am really interested in your live demo , interpolation, and outlier points filtering. Could you share the code to me? My email is zhoufancheng@gmail.com. Thank you!
